### DISCUSSION TITLE
Creating a audio file using PCM data which is converted to bytes, getting audio played with a different pitch and slowed

### DISCUSSION THREAD

**Discussion by XavierSeh** [community/community#122593] [Programming Help]
### Select Topic Area

General

### Body

I am using a voice call functionality in which i am extracting the audio data, the audio data i get is in PCM format which i converts into bytes array and then write it to a .wav file and then download it. On playing that file the voice in it is actually playing like it is getting played in slow motion. Below is the code snippet i am using to do this:

const audioData = {
track: null,
channels: []
}

this is how i am storing the PCM data:
const currentChannelData = buffer.getChannelData(channel);
console.log("PCM data in channel", channel, currentChannelData);
audioData.channels.push(Array.from(currentChannelData));
console.log(buffer.sampleRate)
sampleRate = buffer.sampleRate;

//on logging this sample rate it is, 48000

// at the end this is how i am converting final PCM data to .wav file:
function saveWavFile() {
    // Concatenate all channels
    const mergedChannels = audioData.channels.flat();
    const wavByteArray = pcmToWav(mergedChannels, sampleRate);

    // Create Blob from Uint8Array
    const wavBlob = new Blob([wavByteArray], { type: 'audio/wav' });
    const url = URL.createObjectURL(wavBlob);
    const a = document.createElement('a');
    document.body.appendChild(a);
    a.style = 'display: none';
    a.href = url;
    a.download = 'recording.wav';
    a.click();
    window.URL.revokeObjectURL(url);
  }

function pcmToWav(channelData, sampleRate) {
    const numFrames = channelData.length;
    const arrayBuffer = new ArrayBuffer(44 + numFrames * 2);
    const view = new DataView(arrayBuffer);

    // Write the WAV container
    writeString(view, 0, 'RIFF');
    view.setUint32(4, 36 + numFrames * 2, true);
    writeString(view, 8, 'WAVE');
    writeString(view, 12, 'fmt ');
    view.setUint32(16, 16, true);
    view.setUint16(20, 1, true); // PCM format
    view.setUint16(22, 1, true); // Mono (1 channel)
    view.setUint32(24, sampleRate, true);
    view.setUint32(28, sampleRate * 2, true); // Byte rate
    view.setUint16(32, 2, true); // Block align
    view.setUint16(34, 16, true); // 16-bit sample size
    writeString(view, 36, 'data');
    view.setUint32(40, numFrames * 2, true);

    // Write the PCM samples
    for (let i = 0; i < numFrames; i++) {
        view.setInt16(44 + i * 2, channelData[i] * 0x7FFF, true);
    }

    // Convert ArrayBuffer to Uint8Array
    const byteArray = new Uint8Array(arrayBuffer);

    return byteArray;
  }
If someone have any idea on how to resolve this please help.

I have tried directly changing the sample rate to 16000, it make the audio to appear even more slow. I also tried to increase sample rate, the speed of audio speech increases but the pitch of it changes very much.

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @XavierSeh !

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Discussions category is solely related to conversations around the [GitHub product Discussions](https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/about-discussions). This question should be in the Programming Help category. I've gone ahead and moved it for you. Good luck!

**Comment by CopperEagle** [comment]
Hey there @XavierSeh 

Nice job correctly mastering the endian nightmare. Generally, the header seems to be correct.

It is hard to tell from afar what is exactly going on I do not know the exact data that you get from the microphone. There are many tiny nasty details (correct datatype, channels, endianness etc.) that quickly make you sound like talking through pudding. Here are some of my thoughts:

1) As far as I know, channel data should interleave. E.g. `-left- -right- -left- -right-`. If you have 48k, one second consists of 96k samples. So, the line I am rising an eyebrow at is the `const mergedChannels = audioData.channels.flat();`. That does a concatenation. First all left, then all right. 

Now interestingly, you hardcode the number of channels to be 1. I'm just gonna assume that's right, as thats typical for microphones in phones. So that shouldn't be your problem. Still, to check if channels may be a problem: Is setting the samplerate to 48k the speed 1.0x? What about setting it to 96k?

2) Pitch could be connected to the above. Bad length, bad sound. *However*, if you get 1.0x speed and the pitch is still too low, it could be a seperate problem. One thing that I am not sure about is what is your data at the start? Do you get the samples as floats? As ints? Maybe there is some conflicting assumptions between getting the data, then translating that to PCM. Honestly, I'm rather puzzled what `pcmToWav` is supposed to do, since you wav stores PCM. 


Ps. Yes, setting the samplerate from 48k to 16k should make your pitch lower: Your data is getting stretched out over more time, reducing frequencies ;)

**Reply by XavierSeh** [reply]
Hey @CopperEagle thank you for the response, here i tried setting sample rate to 96k and the audio speed did sounded at 1.0x thanks for that. But the problem is that the audio is sounding like a robotic audio, do you have any idea how it could be resolved?
Also the the incoming data i.e. CurrentchannelData it is in float32array format which i am converting to uint8array and then trying to download a .wav file to listen to it if the audio incoming is in correct format.
The reason why i am doing this is that i have server that transcribes audio bytes, so in order to make sure i am sending correct data i was trying to record and listen to a .wav file. Also uint8array can be sent to the server that is why i am first convert it into that first.

https://github.com/community/community/assets/144888233/3a984d63-6105-49c0-a9fd-1e36c1d8d781

This above is the audio file recorded.

**Reply by CopperEagle** [reply]
Hey @XavierSeh

Okay, so we have twice the data than expected. I still think your channel settings are alright, assuming you only have one channel from the mic. Just make sure the audioData.channel has one subarray.

The problem is likely a data issue. Let me recap. You get `float32` from `getChannelData`. You merge your arrays. Then you then call `pcmToWav` essentially with that data. You tell the header "hey, my data is 16 bits". Then, you do `view.setInt16`, which does conversion but not what you want. You only get -1, 0 or 1. There are two ways of dealing with this.

1)  Change your data before storing it, mapping the floats to ints. That is, you map from `[-1, 1]` to the `int16`. You can get the code from [here](https://stackoverflow.com/questions/62874963/audio-float32-to-int16-conversion-in-javascript), from the guy who asks the question.

2) The Mr. Premium - and easier - way is changing the header, instead of the data. So you specify that your data is `float32`. It is premium because you 2x your filesize with potentially little impact on quality, depending on the mic. The fallowing is a draft for that updated header. I haven't tested it, so please doublecheck. Changed lines are commented.

```js
// Write the WAV container
writeString(view, 0, 'RIFF');
view.setUint32(4, 36 + numFrames * 4, true); // Change file size
writeString(view, 8, 'WAVE');
writeString(view, 12, 'fmt ');
view.setUint32(16, 16, true);
view.setUint16(20, 1, true); // PCM format (unchanged)
view.setUint16(22, 1, true); // Mono (1 channel, unchanged)
view.setUint32(24, sampleRate, true);
view.setUint32(28, sampleRate * 4, true); // Set byte rate for 32-bit float
view.setUint16(32, 4, true); // Update block align for 32-bit float
view.setUint16(34, 32, true); // Change to 32-bit float sample size
writeString(view, 36, 'data');
view.setUint32(40, numFrames * 4, true); // Update to actual data size
```

Then, you use `view.setFloat32(...)`, instead of `view.setInt16` for setting the data.


Ps. Silly me. In the first answer, I did not see that the name of the function you defined litterlly was `pcmToWav`. Thought somehow you use a library...
